### PR TITLE
Add Damage per Second via lockout time

### PR DESCRIPTION
**Pull request by Groovman**
Infos here: https://yomisolov-geek.com/%e7%a1%ac%e7%9b%b4%e3%81%a8%e8%a1%8c%e5%8b%95%e3%81%ae%e6%9c%80%e9%81%a9%e5%8c%96%e3%81%ab%e3%81%a4%e3%81%84%e3%81%a6/

From the formula we get Lockout time/turn, we have expected damage per turn. We do expected damage/turn divided by lockout time seconds/turn and get expected damage /seconds so damage per seconds. Yay!

### CODE REVIEW

File: src/global_logic.js [hunk 1]
```diff
@@ -933,6 +933,26 @@ module.exports.calcBasedOneSummon = function (summonind, prof, buff, totals) {
...
+            // Ougi Lockout
+            expectedCycleLockoutTime += 1.0 + 0.35;
+            // Chainburst lockout
+            if (buff["chainNumber"] == 2) {
+                expectedCycleLockoutTime += 5.0 / buff["chainNumber"];
+            } else if (buff["chainNumber"] == 3) {
+                expectedCycleLockoutTime += 7.0 / buff["chainNumber"];
+            } else if (buff["chainNumber"] >= 4) {
+                expectedCycleLockoutTime += 9.0 / buff["chainNumber"] == 4 ? buff["chainNumber"] : 4;
+            }
```
Comment (kei-gbf) [inline src/global_logic.js]: Can normalize the logic to formula? and Math.min for over chain
```js
let chainNumber = Math.min(4, buff["chainNumber"]);
if (chainNumber >= 2) {
    expectedCycleLockoutTime += (chainNumber * 2 + 1.0) / chainNumber;
}
```
- [x] Over chain ... ok
- [x] Skip chainNumber 1 (no chains)
- [x] Skip invalid case 0 or Negative chainNumver

Comment (kei-gbf) [reply src/global_logic.js]: or should be function  getChainLockoutTime(chainNumber)
unit test can check chainNumer -1 .. 5 cases

**Comment by kei-gbf** [comment]
when copy/paste code It's time to make a function
generateHaisuiData uses "chainNumber" but
calcBasedOneSummon uses "newChainNumber"

```js
function calcChainLockoutTime(actualChainNumber) {
    let chainNumber = Math.min(4, actualChainNumber);
    if (chainNumber < 2)
        return 0; // no chain or invalid chain number
    return (chainNumber * 2 + 1.0) / chainNumber:
}
module.exports.calcChainLockoutTime = calcChainLockoutTime;
```

no need to care variable name conflict in the function scope,
no need to search/modify multiple place for maintenance duplicated codes.
and make it possible validate the function behavior by unit test.

----
- if (chainNumber < 2)  for more strict code (optional topic)
  - [ ] no chain (chainNumber: 1) is ok valid input number
  - [ ] over chain ... Motocal UI allow chain number bitween 1..4
    - chain number 5 but program can't know who does multi C.A.?
  - [ ] invalid numbers (e.g. -1 negative number) are not
        the function should report error/warning that can be UI's bugs.
    - [x] float number like 1.5  ... UI does not allow manual input now
    - [x] non number format string ... UI does not allow manual input now
      - [ ] for future safety, reject those invalid inputs

----
and the unit tests (not tested this code, sorry if it had typo or errors)

```js
// import calcChainLockoutTime in global_logic.test.js

desctibe("calcChainLockoutTIme", () => {
    test("legal chain number", () => {
        expect(calcChainLockoutTime(2)).toBe(5);
        expect(calcChainLockoutTime(3)).toBe(7);
        expect(calcChainLockoutTime(4)).toBe(9);
    });
    test("no chain should return 0", () => {
        expect(calcChainLockoutTime(1)).toBe(0);
    });    
    test("over chain are equal to 4chain", () => {
        expect(calcChainLockoutTime(5)).toBe(9);
        expect(calcChainLockoutTime(6)).toBe(9);
        expect(calcChainLockoutTime(7)).toBe(9);
        expect(calcChainLockoutTime(8)).toBe(9);
        expect(calcChainLockoutTime(9)).toBe(9);
    });
    test("invalid input chainNumber", () => {
        // this implementation returns 0 ignore the invalid cases.
        expect(calcChainLockoutTime(0)).toBe(0);
        expect(calcChainLockoutTime(-1)).toBe(0);
    });
});
```

**Comment by kei-gbf** [comment]
Tested
- [x] The lockout time is affected by DA/TA and Ougi gage up
  - [ ] no normal attack chara ? (Lyria, Sakura)
    - how about implement it as ougi gage up 100% down ?
- [x] There was no chance to set over chain
  - gage 200% chara has effect to the lockout time ?

File: src/global_const.js [hunk 2]
```diff
@@ -1152,7 +1152,7 @@ var raceTypes = {
 var sexTypes = {
     "female": "女",
     "male": "男",
-    "other": "不詳",
+    "other": "不明",
```
Comment (kei-gbf) [inline src/global_const.js]: other

「不詳」in game, Can we solve this in another separate PR? it's off topic and I met same error.

Comment (Groovman) [reply src/global_const.js]: Just merge your fix in here.

Comment (Groovman) [reply src/global_const.js]: Did it myself

**Comment by Groovman** [comment]
>   * [ ]  no normal attack chara ? (Lyria, Sakura)

No atk should still result in base + 1 atk, but they need to be implemented fully the function should correct itself if they are.

**Comment by Groovman** [comment]
Please merge, it is tested for months now.

**Comment by kagekiyo7** [comment]
This PR is not considered about the frequency of a chain burst occurring.
Lockout time is mostly based on number of chain burst.
Here is my bad sample code that considered it. It is calculate `100T damage / 100T lockoutTime` that based on ougi gage rule of granblue.
https://github.com/kagekiyo7/motocal/blob/Add-new-calc-lockoutTime/src/newcalc.js

and new verification.
https://note.com/donnguri_gbf/n/n277cc26533ae
https://note.com/donnguri_gbf/n/n1c52c6acff4b
https://twitter.com/JeffreyManson/status/869847039013670912

**Comment by kagekiyo7** [comment]
I did refactoring (no test)
https://github.com/kagekiyo7/motocal/blob/Add-new-calc-lockoutTime/src/newcalc.js

**Comment by Groovman** [comment]
Sorry I am busy this event because I also have work.

@kagekiyo7 If you want to extend it feel free to do it and push  to my branch and merge it to the motocal master when you feel it is good :D

**Pull request merged**